### ISSUE TITLE
Behaviour change/Question: Failed inference on `dict.get("key", [])`

### ISSUE THREAD

**Issue by pvandyken** [microsoft/pyright#5781]
**Describe the bug**
This issue is related to #5364, where `TypedDict.get("key", [])` led to a partially unknown argument. That issue was resolved by interpreting the empty list as `list[Any]`, in order to make it consistent with the previous behaviour of a regular dict.

Now regular dicts are giving unknown arguments: in `dict.get("key", [])`, the list is interpreted as `list[Unknown]`. 

So I really have two problems: one, we once again have inconsistency between `TypedDict` and `dict`. One behaviour or the other should apply to both.

Two, admittedly, it's probably more correct to infer the empty list as `list[Unknown]`. The problem with this behaviour is it makes the following pattern give lint errors:

```py
my_dict.get("key", {}).get("subkey", {})
```

I frequently use this pattern to access potentially nested dicts. There's other convenience patterns that become disrupted as well, for instance:

```py
[g for g in my_dict.get("my_list", [])]
```

In the event that `list[Unknown]` is selected as the correct behaviour, is there a way to make the above patterns work without `# type: ignore` or any equally elegant alternatives?

**Code or Screenshots**
If possible, provide a minimal, self-contained code sample (surrounded by triple back ticks) to demonstrate the issue. The code should define or import all referenced symbols.

```python
# test.py
from __future__ import annotations
from typing import Any, TypedDict

class typed_dict(TypedDict, total=False):
    dict_key: dict[str, Any]
    list_key: list[str]

def func(
    tdict: typed_dict,
    reg_dict: dict[str, dict[str, Any]],
):
    tdict.get("dict_key", [])
    # No error

    tdict.get("list_key", [])
    # No error

    reg_dict.get("key", [])
    #                   ~~
    # Argument type is "list[Unknown]"
```

Output:

```
/path/to/test.py
  /path/to/test.py:19:25 - error: Argument type is partially unknown
  Argument corresponds to parameter "__default" in function "get"
  Argument type is "list[Unknown]" (reportUnknownArgumentType)
1 error, 0 warnings, 0 informations
```

**VS Code extension or command-line**
pyright==1.1.323

**Comment by erictraut** [comment]
This change was due to a recent modification to the definition of `dict.get` in the typeshed stubs. I understand the reason behind the change, but I think it was a relatively risk one. I even called it out in the [pyright 1.1.323 release notes](https://github.com/microsoft/pyright/releases/tag/1.1.323).

> Enhancement: Updated typeshed to the latest. This update includes some changes to the two-parameter form of the get method for dict and TypedDict that may result in visible type evaluation changes in some code bases.

This also underscores the problems related to hard-coding type information in type checkers. In the case of `TypedDict`, we don't have a choice but to use hard-coded logic to synthesize methods like `get`. This unfortunately means it can easily get out of sync with typeshed. I don't have a good solution to offer here.

I'm reluctant to change the synthesized TypedDict `get` method (again) because the justification for doing so is pretty weak. I agree with you that consistency between TypedDict and dict would be preferable, but I'm not going to lose much sleep over this one. Making the change wouldn't provide any other value from a type checking perspective, and it would create churn for existing pyright users. So for now, I'm not going to make any change here. I'm open to revisiting this decision in the future if there's a compelling argument.

To answer your other question, I don't see a way to make that coding pattern work without generating an `Unknown` error, at least not with the latest typeshed stubs.

**Comment by pvandyken** [comment]
Thanks for the detailed response. As it turns out, I was a bit quick on the draw with this issue: so long as the same container type is used as the default, inference works just fine (note in the example above, I used a list as default rather than a dict). My confusion arose initially because the dicts I was working with were dicts of `TypeDict`s, which count as different container types. So I was using a plain `dict` as a default to `get()` when I should have used an instance of `TypedDict`.
